File: java/src/PpmModel.java
/* 
 * Reference arithmetic coding
 * Copyright (c) Project Nayuki
 * 
 * https://www.nayuki.io/page/reference-arithmetic-coding
 * https://github.com/nayuki/Reference-arithmetic-coding
 */


import java.util.Arrays;

final class PpmModel {
	
	/*---- Fields ----*/
	
	public final int modelOrder;
	
	private final int symbolLimit;
	final int escapeSymbol;
	
	public final Context rootContext;
	public final FrequencyTable orderMinus1Freqs;
	
	
	
	/*---- Constructors ----*/
	
	public PpmModel(int order, int numSymbols) {
		if (order < -1 || numSymbols <= 1)
			throw new IllegalArgumentException();
		this.modelOrder = order;
		this.symbolLimit = numSymbols + 1;
		this.escapeSymbol = numSymbols;
		
		if (order >= 0) {
			rootContext = new Context(symbolLimit);
		} else
			rootContext = null;
		orderMinus1Freqs = new FlatFrequencyTable(symbolLimit);
	}
	
	
	
	/*---- Methods ----*/
	
	public void incrementContexts(int[] history, int symbol) {
		if (modelOrder == -1)
			return;
		if (history.length > modelOrder || symbol < 0 || symbol >= symbolLimit)
			throw new IllegalArgumentException();
		
		Context ctx = rootContext;
		ctx.frequencies.increment(symbol);
		for (int sym : history) {
			Context[] subctxs = ctx.getSubcontexts();
			
			if (subctxs[sym] == null) {
				subctxs[sym] = new Context(symbolLimit);
			}
			ctx = subctxs[sym];
			ctx.frequencies.increment(symbol);
		}
	}
	
	public int[] addToHistory(int[] history, int symbol) {
		if (modelOrder >= 1) {
			// Prepend current symbol, dropping oldest symbol if necessary
			if (history.length < modelOrder)
				history = Arrays.copyOf(history, history.length + 1);
			System.arraycopy(history, 0, history, 1, history.length - 1);
			history[0] = symbol;
		}
		return history;
	}
	
	
	/*---- Helper structure ----*/
	
	public static final class Context {
		
		public final FrequencyTable frequencies;
		
		private Context[] subcontexts;
		
		public Context(int symbols) {
			frequencies = new SimpleFrequencyTable(new int[symbols]);
			frequencies.increment(symbols - 1);
		}
		
		public Context[] getSubcontexts() {
			if (subcontexts == null) {
				subcontexts = new Context[frequencies.getSymbolLimit()];
			}
			return subcontexts;
		}
	}
	
}

File: python/ppmmodel.py
# 
# Reference arithmetic coding
# 
# Copyright (c) Project Nayuki
# MIT License. See readme file.
# https://www.nayuki.io/page/reference-arithmetic-coding
# 

import arithmeticcoding


class PpmModel:
	
	def __init__(self, order, symbollimit, escapesymbol):
		if not ((order >= -1) and (0 <= escapesymbol < symbollimit)):
			raise ValueError()
		self.model_order = order
		self.symbol_limit = symbollimit
		self.escape_symbol = escapesymbol
		
		if order >= 0:
			self.root_context = PpmModel.Context(symbollimit, order >= 1)
			self.root_context.frequencies.increment(escapesymbol)
		else:
			self.root_context = None
		self.order_minus1_freqs = arithmeticcoding.FlatFrequencyTable(symbollimit)
	
	
	def increment_contexts(self, history, symbol):
		if self.model_order == -1:
			return
		if not ((len(history) <= self.model_order) and (0 <= symbol < self.symbol_limit)):
			raise ValueError()
		
		ctx = self.root_context
		ctx.frequencies.increment(symbol)
		for (i, sym) in enumerate(history):
			subctxs = ctx.subcontexts
			assert subctxs is not None
			
			if subctxs[sym] is None:
				subctxs[sym] = PpmModel.Context(self.symbol_limit, i + 1 < self.model_order)
				subctxs[sym].frequencies.increment(self.escape_symbol)
			ctx = subctxs[sym]
			ctx.frequencies.increment(symbol)
	
	
	
	# Helper structure
	class Context:
		
		def __init__(self, symbols, hassubctx):
			self.frequencies = arithmeticcoding.SimpleFrequencyTable([0] * symbols)
			self.subcontexts = ([None] * symbols) if hassubctx else None
